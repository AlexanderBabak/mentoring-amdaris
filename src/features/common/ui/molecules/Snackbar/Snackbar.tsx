import React, { FC, ReactNode } from "react";
import {
  Alert,
  AlertColor,
  AlertTitle,
  Snackbar as MuiSnackbar,
  SnackbarProps as MuiSnackbarProps,
} from "@mui/material";

export interface SnackbarProps extends Omit<MuiSnackbarProps, "onClose"> {
  onClose: (event: Event | React.SyntheticEvent<any, Event>) => void;
  alertSeverity?: AlertColor;
  alertVariant?: "filled" | "standard" | "outlined";
  alertTitle?: string;
  alertContent: ReactNode;
  alertAction?: ReactNode;
}

const Snackbar: FC<SnackbarProps> = ({
  open,
  onClose,
  alertSeverity,
  autoHideDuration = 10000,
  anchorOrigin = {
    horizontal: "right",
    vertical: "bottom",
  },
  alertVariant = "filled",
  alertTitle,
  alertContent,
  alertAction,
  ClickAwayListenerProps,
  ...props
}) => {
  return (
    <MuiSnackbar
      data-testid="snackbar"
      open={open}
      autoHideDuration={autoHideDuration}
      onClose={onClose}
      anchorOrigin={anchorOrigin}
      ClickAwayListenerProps={ClickAwayListenerProps}
      {...props}
    >
      <Alert
        data-testid={alertSeverity}
        onClose={onClose}
        severity={alertSeverity}
        sx={{ color: (theme) => theme.palette.common.white, alignItems: "center" }}
        variant={alertVariant}
        action={alertAction}
      >
        {!!alertTitle && <AlertTitle sx={{ fontWeight: 700, mb: 0 }}>{alertTitle}</AlertTitle>}

        {alertContent}
      </Alert>
    </MuiSnackbar>
  );
};

export default Snackbar;
