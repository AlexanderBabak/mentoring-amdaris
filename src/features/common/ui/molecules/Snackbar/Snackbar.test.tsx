import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Snackbar, { SnackbarProps } from "./Snackbar";

const onCloseMock = jest.fn();
const baseProps: SnackbarProps = {
  open: true,
  onClose: onCloseMock,
  alertSeverity: "success",
  alertTitle: "some alert title",
  alertContent: "some alert content",
};

describe("Snackbar", () => {
  it("should close as expected", async () => {
    const closeProps = {
      ...baseProps,
      open: false,
    };

    render(<Snackbar {...closeProps} />);

    expect(screen.queryByRole("alert")).not.toBeInTheDocument();
  });

  it("should call close callback as expected", async () => {
    const user = userEvent.setup();

    render(<Snackbar {...baseProps} />);

    expect(screen.getByRole("alert")).toBeInTheDocument();

    const closeButton = screen.getByRole("button", {
      name: /close/i,
    });

    await user.click(closeButton);
    expect(onCloseMock).toHaveBeenCalledTimes(1);
  });

  it("should be rendered as expected", async () => {
    const { baseElement } = render(<Snackbar {...baseProps} />);

    expect(baseElement).toMatchSnapshot();
  });
});
