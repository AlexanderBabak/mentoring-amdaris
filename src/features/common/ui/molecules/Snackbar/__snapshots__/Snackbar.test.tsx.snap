// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snackbar should be rendered as expected 1`] = `
.emotion-0 {
  z-index: 1400;
  position: fixed;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  left: 8px;
  right: 8px;
  -webkit-box-pack: end;
  -ms-flex-pack: end;
  -webkit-justify-content: flex-end;
  justify-content: flex-end;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  bottom: 8px;
}

@media (min-width:600px) {
  .emotion-0 {
    bottom: 24px;
    right: 24px;
    left: auto;
  }
}

.emotion-1 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  border-radius: 4px;
  box-shadow: none;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 0.875rem;
  line-height: 1.43;
  letter-spacing: 0.01071em;
  background-color: transparent;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 6px 16px;
  font-weight: 500;
  background-color: #2e7d32;
  color: #fff;
  color: #fff;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-2 {
  margin-right: 12px;
  padding: 7px 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  font-size: 22px;
  opacity: 0.9;
}

.emotion-3 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: inherit;
}

.emotion-4 {
  padding: 8px 0;
  min-width: 0;
  overflow: auto;
}

.emotion-5 {
  margin: 0;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.5;
  letter-spacing: 0.00938em;
  margin-bottom: 0.35em;
  font-weight: 500;
  margin-top: -2px;
  font-weight: 700;
  margin-bottom: 0px;
}

.emotion-6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  padding: 4px 0 0 16px;
  margin-left: auto;
  margin-right: -8px;
}

.emotion-7 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0, 0, 0, 0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: inherit;
  padding: 5px;
  font-size: 1.125rem;
}

.emotion-7::-moz-focus-inner {
  border-style: none;
}

.emotion-7.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-7 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-7:hover {
  background-color: rgba(0, 0, 0, 0.04);
}

@media (hover: none) {
  .emotion-7:hover {
    background-color: transparent;
  }
}

.emotion-7.Mui-disabled {
  background-color: transparent;
  color: rgba(0, 0, 0, 0.26);
}

.emotion-8 {
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  width: 1em;
  height: 1em;
  display: inline-block;
  fill: currentColor;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: fill 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  font-size: 1.25rem;
}

.emotion-9 {
  overflow: hidden;
  pointer-events: none;
  position: absolute;
  z-index: 0;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  border-radius: inherit;
}

<body>
  <div>
    <div
      class="MuiSnackbar-root MuiSnackbar-anchorOriginBottomRight emotion-0"
      data-testid="snackbar"
      role="presentation"
    >
      <div
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-colorSuccess MuiAlert-filledSuccess MuiAlert-filled emotion-1"
        data-testid="success"
        direction="up"
        role="alert"
        style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      >
        <div
          class="MuiAlert-icon emotion-2"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit emotion-3"
            data-testid="SuccessOutlinedIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M20,12A8,8 0 0,1 12,20A8,8 0 0,1 4,12A8,8 0 0,1 12,4C12.76,4 13.5,4.11 14.2, 4.31L15.77,2.74C14.61,2.26 13.34,2 12,2A10,10 0 0,0 2,12A10,10 0 0,0 12,22A10,10 0 0, 0 22,12M7.91,10.08L6.5,11.5L11,16L21,6L19.59,4.58L11,13.17L7.91,10.08Z"
            />
          </svg>
        </div>
        <div
          class="MuiAlert-message emotion-4"
        >
          <div
            class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiAlertTitle-root emotion-5"
          >
            some alert title
          </div>
          some alert content
        </div>
        <div
          class="MuiAlert-action emotion-6"
        >
          <button
            aria-label="Close"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall emotion-7"
            tabindex="0"
            title="Close"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall emotion-8"
              data-testid="CloseIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root emotion-9"
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</body>
`;
